File: src/main/java/me/lokka30/levelledmobs/util/SpigotUtils.java
package me.lokka30.levelledmobs.util;

import java.util.List;

import me.lokka30.levelledmobs.LevelledMobs;
import me.lokka30.levelledmobs.wrappers.LivingEntityWrapper;
import me.lokka30.levelledmobs.result.NametagResult;
import net.md_5.bungee.api.chat.ClickEvent;
import net.md_5.bungee.api.chat.HoverEvent;
import net.md_5.bungee.api.chat.TextComponent;
import net.md_5.bungee.api.chat.hover.content.Text;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Entity;
import org.bukkit.entity.LivingEntity;
import org.bukkit.entity.Player;
import org.bukkit.entity.Projectile;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.inventory.meta.ItemMeta;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Provides function for APIs that are deprecated in Paper but required for use in Spigot
 *
 * @author stumper66
 * @since 3.3.0
 */
@SuppressWarnings("deprecation")
public class SpigotUtils {

    public static void sendHyperlink(final @NotNull CommandSender sender, final String message,
        final String url) {

        final TextComponent component = new TextComponent(message);
        component.setClickEvent(new ClickEvent(ClickEvent.Action.OPEN_URL, url));
        component.setHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, new Text(url)));
        sender.sendMessage(component);
    }

    public static void updateItemMetaLore(final @NotNull ItemMeta meta,
        final @Nullable List<String> lore) {
        if (lore == null) {
            meta.setLore(null);
        } else {
            meta.setLore(Utils.colorizeAllInList(lore));
        }
    }

    public static void updateItemDisplayName(final @NotNull ItemMeta meta,
        final @Nullable String displayName) {
        meta.setDisplayName(displayName);
    }

    @NotNull public static String getPlayerDisplayName(final @Nullable Player player) {
        if (player == null) {
            return "";
        }
        return player.getDisplayName();
    }

    @Nullable public static LivingEntityWrapper getPlayersKiller(@NotNull final PlayerDeathEvent event,
        final LevelledMobs main) {
        if (event.getDeathMessage() == null) {
            return null;
        }

        final EntityDamageEvent entityDamageEvent = event.getEntity().getLastDamageCause();
        if (entityDamageEvent == null || entityDamageEvent.isCancelled()
            || !(entityDamageEvent instanceof EntityDamageByEntityEvent)) {
            return null;
        }

        final Entity damager = ((EntityDamageByEntityEvent) entityDamageEvent).getDamager();
        LivingEntity killer = null;

        if (damager instanceof final Projectile projectile) {
            if (projectile.getShooter() instanceof LivingEntity) {
                killer = (LivingEntity) projectile.getShooter();
            }
        } else if (damager instanceof LivingEntity) {
            killer = (LivingEntity) damager;
        }

        if (killer == null || Utils.isNullOrEmpty(killer.getName()) || killer instanceof Player) {
            return null;
        }

        final LivingEntityWrapper lmKiller = LivingEntityWrapper.getInstance(killer, main);
        if (!lmKiller.isLevelled()) {
            return lmKiller;
        }

        final NametagResult nametagResult = main.levelManager.getNametag(lmKiller, true, true);
        String deathMessage = nametagResult.getNametagNonNull()
                .replace("%player%", event.getEntity().getName());
        if (Utils.isNullOrEmpty(deathMessage) || "disabled".equalsIgnoreCase(deathMessage)) {
            return lmKiller;
        }

        if (nametagResult.hadCustomDeathMessage()){
            String nametag = nametagResult.getNametagNonNull();
            if (nametag.contains("{DisplayName}")){
                nametag = nametag.replace("{DisplayName}", main.levelManager.replaceStringPlaceholders(
                        nametagResult.getcustomDeathMessage(), lmKiller, false, null, false));
            }
            event.setDeathMessage(
                    MessageUtils.colorizeAll(nametag.replace("%player%", event.getEntity().getName())));
        }
        else {
            if (deathMessage.contains("{DisplayName}")){
                deathMessage = deathMessage.replace("{DisplayName}", Utils.capitalize(lmKiller.getNameIfBaby()));
            }

            event.setDeathMessage(
                    MessageUtils.colorizeAll(Utils.replaceEx(event.getDeathMessage(), killer.getName(), deathMessage)));
        }

        return lmKiller;
    }
}

File: levelledmobs-plugin/src/main/kotlin/io/github/arcaneplugins/levelledmobs/util/SpigotUtils.kt
@file:Suppress("DEPRECATION")

package io.github.arcaneplugins.levelledmobs.util

import io.github.arcaneplugins.levelledmobs.LevelledMobs
import io.github.arcaneplugins.levelledmobs.util.MessageUtils.colorizeAll
import io.github.arcaneplugins.levelledmobs.util.Utils.capitalize
import io.github.arcaneplugins.levelledmobs.util.Utils.colorizeAllInList
import io.github.arcaneplugins.levelledmobs.util.Utils.replaceEx
import io.github.arcaneplugins.levelledmobs.wrappers.LivingEntityWrapper
import net.md_5.bungee.api.chat.ClickEvent
import net.md_5.bungee.api.chat.HoverEvent
import net.md_5.bungee.api.chat.TextComponent
import net.md_5.bungee.api.chat.hover.content.Text
import org.bukkit.command.CommandSender
import org.bukkit.entity.LivingEntity
import org.bukkit.entity.Player
import org.bukkit.entity.Projectile
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.inventory.meta.ItemMeta

/**
 * Provides function for APIs that are deprecated in Paper but required for use in Spigot
 *
 * @author stumper66
 * @since 3.3.0
 */
object SpigotUtils {
    fun sendHyperlink(
        sender: CommandSender,
        message: String?,
        url: String?
    ) {
        val component = TextComponent(message)
        component.clickEvent =
            ClickEvent(ClickEvent.Action.OPEN_URL, url)
        component.hoverEvent =
            HoverEvent(HoverEvent.Action.SHOW_TEXT, Text(url))
        sender.sendMessage(component)
    }

    fun updateItemMetaLore(
        meta: ItemMeta,
        lore: MutableList<String>?
    ) {
        if (lore == null) {
            meta.lore = null
        } else {
            meta.lore = colorizeAllInList(lore)
        }
    }

    fun updateItemDisplayName(
        meta: ItemMeta,
        displayName: String?
    ) {
        meta.setDisplayName(displayName)
    }

    fun getPlayerDisplayName(
        player: Player?
    ): String {
        if (player == null) {
            return ""
        }
        return player.displayName
    }

    fun getPlayersKiller(
        event: PlayerDeathEvent
    ): LivingEntityWrapper? {
        if (event.deathMessage == null) {
            return null
        }

        val entityDamageEvent = event.entity.lastDamageCause
        if (entityDamageEvent == null || entityDamageEvent.isCancelled
            || entityDamageEvent !is EntityDamageByEntityEvent
        ) {
            return null
        }

        val damager = entityDamageEvent.damager
        var killer: LivingEntity? = null

        if (damager is Projectile) {
            if (damager.shooter is LivingEntity) {
                killer = damager.shooter as LivingEntity?
            }
        } else if (damager is LivingEntity) {
            killer = damager
        }

        if (killer == null || killer.name.isEmpty() || killer is Player) {
            return null
        }

        val lmKiller = LivingEntityWrapper.getInstance(killer)
        lmKiller.associatedPlayer = event.entity
        if (!lmKiller.isLevelled) {
            return lmKiller
        }

        val nametagResult = LevelledMobs.instance.levelManager.getNametag(lmKiller,
            isDeathNametag = true,
            preserveMobName = true
        )
        var deathMessage = nametagResult.nametagNonNull
            .replace("%player%", event.entity.name)
        if (deathMessage.isEmpty() || "disabled".equals(deathMessage, ignoreCase = true)) {
            return lmKiller
        }

        if (nametagResult.hadCustomDeathMessage) {
            var nametag = nametagResult.nametagNonNull
            if (nametag.contains("{DisplayName}")) {
                nametag = nametag.replace(
                    "{DisplayName}", LevelledMobs.instance.levelManager.replaceStringPlaceholders(
                        nametagResult.customDeathMessage!!, lmKiller, false, null, false
                    )
                )
            }
            event.deathMessage = colorizeAll(nametag.replace("%player%", event.entity.name))
        } else {
            if (deathMessage.contains("{DisplayName}")) {
                deathMessage = deathMessage.replace("{DisplayName}", capitalize(lmKiller.nameIfBaby))
            }

            event.deathMessage = colorizeAll(replaceEx(event.deathMessage!!, killer.name, deathMessage))
        }

        return lmKiller
    }
}